"error: linking with `cc` failed: exit code: 1" regression in release build with const generics

Const generics and release mode and Hc128Rng::seed_from_u64 and rustc 1.45.0-nightly (4bd32c980 2020-05-29) or later cause cc linking error.

To avoid the error, do any of:
- use rustc 1.45.0-nightly (a74d1862d 2020-05-14) or earlier
- remove the `#![feature(const_generics)]`
- remove the `Hc128Rng::seed_from_u64(42)` line
- compile in debug mode

I'm guessing that Hc128Rng::seed_from_u64 is not the only thing that will trigger this, but I have not been able to find any other examples.


### Source code

```rust
#![feature(const_generics)]
use rand::SeedableRng;
use rand_hc::Hc128Rng;

fn main() {
    let _ = Hc128Rng::seed_from_u64(42);
}
```
### Dependancies
```
rand = "0.7.3"
rand_hc = "0.2.0"
```

### Error:
```
error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-m64" "-L" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/home/isaac/go/src/github.com/is8ac/cc_err_demo/target/release/deps/cc_err_demo-949370c0c3e7c5b4.cc_err_demo.5gfojped-cgu.0.rcgu.o" "/home/isaac/go/src/github.com/is8ac/cc_err_demo/target/release/deps/cc_err_demo-949370c0c3e7c5b4.cc_err_demo.5gfojped-cgu.1.rcgu.o" "/home/isaac/go/src/github.com/is8ac/cc_err_demo/target/release/deps/cc_err_demo-949370c0c3e7c5b4.cc_err_demo.5gfojped-cgu.2.rcgu.o" "/home/isaac/go/src/github.com/is8ac/cc_err_demo/target/release/deps/cc_err_demo-949370c0c3e7c5b4.cc_err_demo.5gfojped-cgu.3.rcgu.o" "/home/isaac/go/src/github.com/is8ac/cc_err_demo/target/release/deps/cc_err_demo-949370c0c3e7c5b4.cc_err_demo.5gfojped-cgu.4.rcgu.o" "-o" "/home/isaac/go/src/github.com/is8ac/cc_err_demo/target/release/deps/cc_err_demo-949370c0c3e7c5b4" "/home/isaac/go/src/github.com/is8ac/cc_err_demo/target/release/deps/cc_err_demo-949370c0c3e7c5b4.3rdszsq7ibbmb3tn.rcgu.o" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro" "-Wl,-znow" "-Wl,-O1" "-nodefaultlibs" "-L" "/home/isaac/go/src/github.com/is8ac/cc_err_demo/target/release/deps" "-L" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/home/isaac/go/src/github.com/is8ac/cc_err_demo/target/release/deps/librand_hc-a144c651cd5437de.rlib" "/home/isaac/go/src/github.com/is8ac/cc_err_demo/target/release/deps/librand-bec2ad1f072a8aa0.rlib" "/home/isaac/go/src/github.com/is8ac/cc_err_demo/target/release/deps/librand_chacha-542ae1a91c9af789.rlib" "/home/isaac/go/src/github.com/is8ac/cc_err_demo/target/release/deps/libppv_lite86-6c3f11be887055bf.rlib" "/home/isaac/go/src/github.com/is8ac/cc_err_demo/target/release/deps/librand_core-2e34a4d1d07e0522.rlib" "/home/isaac/go/src/github.com/is8ac/cc_err_demo/target/release/deps/libgetrandom-f56bf65617cc4079.rlib" "/home/isaac/go/src/github.com/is8ac/cc_err_demo/target/release/deps/liblibc-dc8395e94d7842f0.rlib" "/home/isaac/go/src/github.com/is8ac/cc_err_demo/target/release/deps/libcfg_if-bd1d23d91bf28674.rlib" "-Wl,--start-group" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-0df9faf88e74ed3b.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-f9e6a5fb2e2e8cbc.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libhashbrown-05b3477ba39b37d9.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-fd43f1008ae3de7b.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libbacktrace-c69fd8a874cc6882.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libbacktrace_sys-4f9ba7dc6471f5d2.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_demangle-a3adee158d70a602.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-36aa33d5b51fea85.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcfg_if-8ffb67928670fecf.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-95ea48ddc44a311b.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-4f5122dbbdcc70ab.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-68f423c59dc48a40.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-54580c7798bf2225.rlib" "-Wl,--end-group" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-e805fea8b8005cb4.rlib" "-Wl,-Bdynamic" "-lc" "-lm" "-lrt" "-lpthread" "-lutil" "-ldl" "-lutil" "-ldl" "-lrt" "-lpthread" "-lgcc_s" "-lc" "-lm" "-lrt" "-lpthread" "-lutil" "-ldl" "-lutil"
  = note: /usr/bin/ld: /home/isaac/go/src/github.com/is8ac/cc_err_demo/target/release/deps/cc_err_demo-949370c0c3e7c5b4.cc_err_demo.5gfojped-cgu.4.rcgu.o: in function `cc_err_demo::main':
          cc_err_demo.5gfojped-cgu.4:(.text._ZN11cc_err_demo4main17h83b85daad37728b4E+0x27): undefined reference to `core::array::<impl core::convert::AsMut<[T]> for [T; N]>::as_mut'
          /usr/bin/ld: /home/isaac/go/src/github.com/is8ac/cc_err_demo/target/release/deps/cc_err_demo-949370c0c3e7c5b4: hidden symbol `_ZN4core5array92_$LT$impl$u20$core..convert..AsMut$LT$$u5b$T$u5d$$GT$$u20$for$u20$$u5b$T$u3b$$u20$N$u5d$$GT$6as_mut17h6128af6739856507E' isn't defined
          /usr/bin/ld: final link failed: bad value
          collect2: error: ld returned 1 exit status
          

error: aborting due to previous error; 1 warning emitted

error: could not compile `cc_err_demo`.

Caused by:
  process didn't exit successfully: `rustc --crate-name cc_err_demo --edition=2018 src/main.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type bin --emit=dep-info,link -C opt-level=3 -Cembed-bitcode=no -C metadata=949370c0c3e7c5b4 -C extra-filename=-949370c0c3e7c5b4 --out-dir /home/isaac/go/src/github.com/is8ac/cc_err_demo/target/release/deps -L dependency=/home/isaac/go/src/github.com/is8ac/cc_err_demo/target/release/deps --extern rand=/home/isaac/go/src/github.com/is8ac/cc_err_demo/target/release/deps/librand-bec2ad1f072a8aa0.rlib --extern rand_hc=/home/isaac/go/src/github.com/is8ac/cc_err_demo/target/release/deps/librand_hc-a144c651cd5437de.rlib` (exit code: 1)
```

### Version
```
rustc 1.46.0-nightly (feb3536eb 2020-06-09)
binary: rustc
commit-hash: feb3536eba10c2e4585d066629598f03d5ddc7c6
commit-date: 2020-06-09
host: x86_64-unknown-linux-gnu
release: 1.46.0-nightly
LLVM version: 10.0
```

Note that on latest nightly, I am still encountering this issue.

(Any time I use the `rand` crate and `SeedableRng` with `#![feature(const_generics)]`, I trigger this.)

```
rustc 1.56.0-nightly (6d64f7f69 2021-08-19)
binary: rustc
commit-hash: 6d64f7f695943541fe12bb960971403f440d7225
commit-date: 2021-08-19
host: x86_64-unknown-linux-gnu
release: 1.56.0-nightly
LLVM version: 12.0.1
```

Error:
```
error: linking with `cc` failed: exit status: 1
  |
  = note: "cc" "-m64" "ice_demo.ice_demo.2f961bda-cgu.0.rcgu.o" "ice_demo.ice_demo.2f961bda-cgu.1.rcgu.o" "ice_demo.ice_demo.2f961bda-cgu.10.rcgu.o" "ice_demo.ice_demo.2f961bda-cgu.11.rcgu.o" "ice_demo.ice_demo.2f961bda-cgu.12.rcgu.o" "ice_demo.ice_demo.2f961bda-cgu.2.rcgu.o" "ice_demo.ice_demo.2f961bda-cgu.3.rcgu.o" "ice_demo.ice_demo.2f961bda-cgu.4.rcgu.o" "ice_demo.ice_demo.2f961bda-cgu.5.rcgu.o" "ice_demo.ice_demo.2f961bda-cgu.6.rcgu.o" "ice_demo.ice_demo.2f961bda-cgu.7.rcgu.o" "ice_demo.ice_demo.2f961bda-cgu.8.rcgu.o" "ice_demo.ice_demo.2f961bda-cgu.9.rcgu.o" "ice_demo.3cmg1zgsfxewpl4s.rcgu.o" "-Wl,--as-needed" "-L" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,--start-group" "-Wl,-Bstatic" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-1dd6c19f0820d51d.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-4695c0f6311791db.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libminiz_oxide-1a282f8b292d9e3f.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libadler-a54ae5159230894d.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libobject-13671d003abcccd5.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libmemchr-7509adad918d4b3b.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libaddr2line-66d0be1f7efaf5ca.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgimli-41443b46a1557b2c.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_detect-60812e99072283ff.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_demangle-943ba0c1e3f87a89.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libhashbrown-0dbc7e011696d844.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-78b2343cc72ff57a.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-65fa19680bfb5ee4.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcfg_if-72cab8079f9b3b1e.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-32d13a5363efe47b.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-9ee1d5d15e6abbeb.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-52d5241975807511.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-9924c22ae1efcf66.rlib" "-Wl,--end-group" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-96219fb718f2f3e8.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-L" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "ice_demo" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro" "-Wl,-znow" "-nodefaultlibs"
  = note: /usr/bin/ld: ice_demo.ice_demo.2f961bda-cgu.12.rcgu.o: in function `ice_demo::Foo::bar':
          ice_demo.2f961bda-cgu.12:(.text._ZN8ice_demo3Foo3bar17h82239ce279805729E+0x1b): undefined reference to `core::array::<impl core::convert::AsMut<[T]> for [T; N]>::as_mut'
          /usr/bin/ld: ice_demo: hidden symbol `_ZN4core5array92_$LT$impl$u20$core..convert..AsMut$LT$$u5b$T$u5d$$GT$$u20$for$u20$$u5b$T$u3b$$u20$N$u5d$$GT$6as_mut17ha4f6eccff6597757E' isn't defined
          /usr/bin/ld: final link failed: bad value
          collect2: error: ld returned 1 exit status
          
  = help: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)

error: aborting due to previous error; 1 warning emitted
```

Sorry to keep bumping. I continue to encounter this issue when I attempt to use `SeedableRng` with `generic_const_exprs` on latest nightly.

<details>
  <summary>Code</summary>

```
#![feature(generic_const_exprs)]
use std::convert::AsMut;
use std::default::Default;

trait Foo: Sized {
    type Baz: Default + AsMut<[u8]>;
    fn bar() {
        Self::Baz::default().as_mut();
    }
}

impl Foo for () {
    type Baz = [u8; 1 * 1];
    //type Baz = [u8; 1];
}

fn main() {
    <() as Foo>::bar();
}
```
</details>

<details>
  <summary>Error</summary>

```
error: linking with `cc` failed: exit status: 1
  |
  = note: "cc" "-m64" "cc_error.cc_error.8de2614d-cgu.0.rcgu.o" "cc_error.cc_error.8de2614d-cgu.1.rcgu.o" "cc_error.cc_error.8de2614d-cgu.10.rcgu.o" "cc_error.cc_error.8de2614d-cgu.2.rcgu.o" "cc_error.cc_error.8de2614d-cgu.3.rcgu.o" "cc_error.cc_error.8de2614d-cgu.4.rcgu.o" "cc_error.cc_error.8de2614d-cgu.5.rcgu.o" "cc_error.cc_error.8de2614d-cgu.6.rcgu.o" "cc_error.cc_error.8de2614d-cgu.7.rcgu.o" "cc_error.cc_error.8de2614d-cgu.8.rcgu.o" "cc_error.cc_error.8de2614d-cgu.9.rcgu.o" "cc_error.1gaqyexc5usivrjm.rcgu.o" "-Wl,--as-needed" "-L" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,--start-group" "-Wl,-Bstatic" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-c46e4336da0ba77b.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-9cb5ef8235cf10de.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libminiz_oxide-c94e3d0069d19330.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libadler-5a532766d1e80b5b.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libobject-d08f0c83526179dc.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libmemchr-d995c6cafdb3d253.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libaddr2line-02ebc04a72bd9433.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgimli-5cbc32ae14534ca9.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_detect-969ebc8623abfcc9.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_demangle-7f98f837d3579544.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libhashbrown-b96a8aea2dd7bf12.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-c7b2792f3bc0d0eb.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-7558dde8464d46ce.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcfg_if-b3d84848ab790cf7.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-f48e57d6d73020ee.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-ffaac76756e3f83b.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-2a6a2797f7a73818.rlib" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-0e3656b1fda5fd7b.rlib" "-Wl,--end-group" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-f98cce041f593917.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-L" "/home/isaac/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "cc_error" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro,-znow" "-nodefaultlibs"
  = note: /usr/bin/ld: cc_error.cc_error.8de2614d-cgu.0.rcgu.o: in function `cc_error::Foo::bar':
          cc_error.8de2614d-cgu.0:(.text._ZN8cc_error3Foo3bar17hc2d30ef1d966820bE+0x1b): undefined reference to `core::array::<impl core::convert::AsMut<[T]> for [T; N]>::as_mut'
          /usr/bin/ld: cc_error: hidden symbol `_ZN4core5array92_$LT$impl$u20$core..convert..AsMut$LT$$u5b$T$u5d$$GT$$u20$for$u20$$u5b$T$u3b$$u20$N$u5d$$GT$6as_mut17hc1c54643f022e16cE' isn't defined
          /usr/bin/ld: final link failed: bad value
          collect2: error: ld returned 1 exit status
          
  = help: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)

error: aborting due to previous error; 1 warning emitted
```

</details>

<details>
  <summary>rustc --version --verbose</summary>

```
rustc 1.59.0-nightly (cfa3fe5af 2021-12-31)
binary: rustc
commit-hash: cfa3fe5af339e724209b25715282adae0c61628f
commit-date: 2021-12-31
host: x86_64-unknown-linux-gnu
release: 1.59.0-nightly
LLVM version: 13.0.0
```
</details>

--- a/compiler/rustc_trait_selection/src/traits/query/normalize.rs
+++ b/compiler/rustc_trait_selection/src/traits/query/normalize.rs
@@ -194,7 +194,7 @@ impl<'cx, 'tcx> FallibleTypeFolder<'tcx> for QueryNormalizer<'cx, 'tcx> {
         // wait to fold the substs.
 
         // Wrap this in a closure so we don't accidentally return from the outer function
-        let mut res = (|| match *ty.kind() {
+        let res = (|| match *ty.kind() {
             // This is really important. While we *can* handle this, this has
             // severe performance implications for large opaque types with
             // late-bound regions. See `issue-88862` benchmark.
@@ -266,7 +266,15 @@ impl<'cx, 'tcx> FallibleTypeFolder<'tcx> for QueryNormalizer<'cx, 'tcx> {
                 debug!("QueryNormalizer: result = {:#?}", result);
                 debug!("QueryNormalizer: obligations = {:#?}", obligations);
                 self.obligations.extend(obligations);
-                Ok(result.normalized_ty)
+
+                let res = result.normalized_ty;
+                // `tcx.normalize_projection_ty` may normalize to a type that still has
+                // unevaluated consts, so keep normalizing here if that's the case.
+                if res != ty && res.has_type_flags(ty::TypeFlags::HAS_CT_PROJECTION) {
+                    Ok(res.try_super_fold_with(self)?)
+                } else {
+                    Ok(res)
+                }
             }
 
             ty::Projection(data) => {
@@ -305,24 +313,26 @@ impl<'cx, 'tcx> FallibleTypeFolder<'tcx> for QueryNormalizer<'cx, 'tcx> {
                 debug!("QueryNormalizer: result = {:#?}", result);
                 debug!("QueryNormalizer: obligations = {:#?}", obligations);
                 self.obligations.extend(obligations);
-                Ok(crate::traits::project::PlaceholderReplacer::replace_placeholders(
+
+                let res = crate::traits::project::PlaceholderReplacer::replace_placeholders(
                     infcx,
                     mapped_regions,
                     mapped_types,
                     mapped_consts,
                     &self.universes,
                     result.normalized_ty,
-                ))
+                );
+                // `tcx.normalize_projection_ty` may normalize to a type that still has
+                // unevaluated consts, so keep normalizing here if that's the case.
+                if res != ty && res.has_type_flags(ty::TypeFlags::HAS_CT_PROJECTION) {
+                    Ok(res.try_super_fold_with(self)?)
+                } else {
+                    Ok(res)
+                }
             }
 
             _ => ty.try_super_fold_with(self),
         })()?;
-
-        // `tcx.normalize_projection_ty` may normalize to a type that still has
-        // unevaluated consts, so keep normalizing here if that's the case.
-        if res != ty && res.has_type_flags(ty::TypeFlags::HAS_CT_PROJECTION) {
-            res = res.try_super_fold_with(self)?;
-        }
 
         self.cache.insert(ty, res);
         Ok(res)

--- a/compiler/rustc_trait_selection/src/traits/query/normalize.rs
+++ b/compiler/rustc_trait_selection/src/traits/query/normalize.rs
@@ -194,7 +194,7 @@ impl<'cx, 'tcx> FallibleTypeFolder<'tcx> for QueryNormalizer<'cx, 'tcx> {
         // wait to fold the substs.
 
         // Wrap this in a closure so we don't accidentally return from the outer function
-        let res = (|| match *ty.kind() {
+        let mut res = (|| match *ty.kind() {
             // This is really important. While we *can* handle this, this has
             // severe performance implications for large opaque types with
             // late-bound regions. See `issue-88862` benchmark.
@@ -317,6 +317,13 @@ impl<'cx, 'tcx> FallibleTypeFolder<'tcx> for QueryNormalizer<'cx, 'tcx> {
 
             _ => ty.try_super_fold_with(self),
         })()?;
+
+        // `tcx.normalize_projection_ty` may normalize to a type that still has
+        // unevaluated consts, so keep normalizing here if that's the case.
+        if res != ty && res.has_type_flags(ty::TypeFlags::HAS_CT_PROJECTION) {
+            res = res.try_super_fold_with(self)?;
+        }
+
         self.cache.insert(ty, res);
         Ok(res)
     }

--- a/src/test/ui/const-generics/generic_const_exprs/issue-100217.rs
+++ b/src/test/ui/const-generics/generic_const_exprs/issue-100217.rs
@@ -1,0 +1,42 @@
+// build-pass
+
+#![allow(incomplete_features)]
+#![feature(generic_const_exprs)]
+
+trait TraitOne {
+    const MY_NUM: usize;
+    type MyErr: std::fmt::Debug;
+
+    fn do_one_stuff(arr: [u8; Self::MY_NUM]) -> Result<(), Self::MyErr>;
+}
+
+trait TraitTwo {
+    fn do_two_stuff();
+}
+
+impl<O: TraitOne> TraitTwo for O
+where
+    [(); Self::MY_NUM]:,
+{
+    fn do_two_stuff() {
+        O::do_one_stuff([5; Self::MY_NUM]).unwrap()
+    }
+}
+
+struct Blargotron;
+
+#[derive(Debug)]
+struct ErrTy<const N: usize>([(); N]);
+
+impl TraitOne for Blargotron {
+    const MY_NUM: usize = 3;
+    type MyErr = ErrTy<{ Self::MY_NUM }>;
+
+    fn do_one_stuff(_arr: [u8; Self::MY_NUM]) -> Result<(), Self::MyErr> {
+        Ok(())
+    }
+}
+
+fn main() {
+    Blargotron::do_two_stuff();
+}

--- a/src/test/ui/const-generics/generic_const_exprs/issue-73298.rs
+++ b/src/test/ui/const-generics/generic_const_exprs/issue-73298.rs
@@ -1,0 +1,23 @@
+// build-pass
+
+#![allow(incomplete_features)]
+#![feature(generic_const_exprs)]
+
+use std::convert::AsMut;
+use std::default::Default;
+
+trait Foo: Sized {
+    type Baz: Default + AsMut<[u8]>;
+    fn bar() {
+        Self::Baz::default().as_mut();
+    }
+}
+
+impl Foo for () {
+    type Baz = [u8; 1 * 1];
+    //type Baz = [u8; 1];
+}
+
+fn main() {
+    <() as Foo>::bar();
+}

--- a/src/test/ui/const-generics/generic_const_exprs/issue-82268.rs
+++ b/src/test/ui/const-generics/generic_const_exprs/issue-82268.rs
@@ -1,0 +1,73 @@
+// build-pass
+
+#![allow(incomplete_features)]
+#![feature(generic_const_exprs)]
+
+trait Collate<Op> {
+    type Pass;
+    type Fail;
+
+    fn collate(self) -> (Self::Pass, Self::Fail);
+}
+
+impl<Op> Collate<Op> for () {
+    type Pass = ();
+    type Fail = ();
+
+    fn collate(self) -> ((), ()) {
+        ((), ())
+    }
+}
+
+trait CollateStep<X, Prev> {
+    type Pass;
+    type Fail;
+    fn collate_step(x: X, prev: Prev) -> (Self::Pass, Self::Fail);
+}
+
+impl<X, P, F> CollateStep<X, (P, F)> for () {
+    type Pass = (X, P);
+    type Fail = F;
+
+    fn collate_step(x: X, (p, f): (P, F)) -> ((X, P), F) {
+        ((x, p), f)
+    }
+}
+
+struct CollateOpImpl<const MASK: u32>;
+trait CollateOpStep {
+    type NextOp;
+    type Apply;
+}
+
+impl<const MASK: u32> CollateOpStep for CollateOpImpl<MASK>
+where
+    CollateOpImpl<{ MASK >> 1 }>: Sized,
+{
+    type NextOp = CollateOpImpl<{ MASK >> 1 }>;
+    type Apply = ();
+}
+
+impl<H, T, Op: CollateOpStep> Collate<Op> for (H, T)
+where
+    T: Collate<Op::NextOp>,
+    Op::Apply: CollateStep<H, (T::Pass, T::Fail)>,
+{
+    type Pass = <Op::Apply as CollateStep<H, (T::Pass, T::Fail)>>::Pass;
+    type Fail = <Op::Apply as CollateStep<H, (T::Pass, T::Fail)>>::Fail;
+
+    fn collate(self) -> (Self::Pass, Self::Fail) {
+        <Op::Apply as CollateStep<H, (T::Pass, T::Fail)>>::collate_step(self.0, self.1.collate())
+    }
+}
+
+fn collate<X, const MASK: u32>(x: X) -> (X::Pass, X::Fail)
+where
+    X: Collate<CollateOpImpl<MASK>>,
+{
+    x.collate()
+}
+
+fn main() {
+    dbg!(collate::<_, 5>(("Hello", (42, ('!', ())))));
+}

--- a/src/test/ui/const-generics/generic_const_exprs/issue-83972.rs
+++ b/src/test/ui/const-generics/generic_const_exprs/issue-83972.rs
@@ -1,0 +1,38 @@
+// build-pass
+
+#![allow(incomplete_features)]
+#![feature(generic_const_exprs)]
+
+pub trait Foo {
+    fn foo(&self);
+}
+
+pub struct FooImpl<const N: usize>;
+impl<const N: usize> Foo for FooImpl<N> {
+    fn foo(&self) {}
+}
+
+pub trait Bar: 'static {
+    type Foo: Foo;
+    fn get() -> &'static Self::Foo;
+}
+
+struct BarImpl;
+impl Bar for BarImpl {
+    type Foo = FooImpl<
+        {
+            { 4 }
+        },
+    >;
+    fn get() -> &'static Self::Foo {
+        &FooImpl
+    }
+}
+
+pub fn boom<B: Bar>() {
+    B::get().foo();
+}
+
+fn main() {
+    boom::<BarImpl>();
+}

--- a/src/test/ui/const-generics/generic_const_exprs/issue-84669.rs
+++ b/src/test/ui/const-generics/generic_const_exprs/issue-84669.rs
@@ -1,0 +1,30 @@
+// build-pass
+
+#![feature(generic_const_exprs)]
+#![allow(incomplete_features)]
+
+trait Foo {
+    type Output;
+
+    fn foo() -> Self::Output;
+}
+
+impl Foo for [u8; 3] {
+    type Output = [u8; 1 + 2];
+
+    fn foo() -> [u8; 3] {
+        [1u8; 3]
+    }
+}
+
+fn bug<const N: usize>()
+where
+    [u8; N]: Foo,
+    <[u8; N] as Foo>::Output: AsRef<[u8]>,
+{
+    <[u8; N]>::foo().as_ref();
+}
+
+fn main() {
+    bug::<3>();
+}

--- a/src/test/ui/const-generics/generic_const_exprs/issue-86710.rs
+++ b/src/test/ui/const-generics/generic_const_exprs/issue-86710.rs
@@ -1,0 +1,73 @@
+// build-pass
+
+#![allow(incomplete_features)]
+#![feature(generic_const_exprs)]
+
+use std::marker::PhantomData;
+
+fn main() {
+    let x = FooImpl::<BarImpl<1>> { phantom: PhantomData };
+    let _ = x.foo::<BarImpl<1>>();
+}
+
+trait Foo<T>
+where
+    T: Bar,
+{
+    fn foo<U>(&self)
+    where
+        T: Operation<U>,
+        <T as Operation<U>>::Output: Bar;
+}
+
+struct FooImpl<T>
+where
+    T: Bar,
+{
+    phantom: PhantomData<T>,
+}
+
+impl<T> Foo<T> for FooImpl<T>
+where
+    T: Bar,
+{
+    fn foo<U>(&self)
+    where
+        T: Operation<U>,
+        <T as Operation<U>>::Output: Bar,
+    {
+        <<T as Operation<U>>::Output as Bar>::error_occurs_here();
+    }
+}
+
+trait Bar {
+    fn error_occurs_here();
+}
+
+struct BarImpl<const N: usize>;
+
+impl<const N: usize> Bar for BarImpl<N> {
+    fn error_occurs_here() {}
+}
+
+trait Operation<Rhs> {
+    type Output;
+}
+
+//// Part-A: This causes error.
+impl<const M: usize, const N: usize> Operation<BarImpl<M>> for BarImpl<N>
+where
+    BarImpl<{ N + M }>: Sized,
+{
+    type Output = BarImpl<{ N + M }>;
+}
+
+//// Part-B: This doesn't cause error.
+// impl<const M: usize, const N: usize> Operation<BarImpl<M>> for BarImpl<N> {
+//     type Output = BarImpl<M>;
+// }
+
+//// Part-C: This also doesn't cause error.
+// impl<const M: usize, const N: usize> Operation<BarImpl<M>> for BarImpl<N> {
+//     type Output = BarImpl<{ M }>;
+// }

